design/thin-edge-core.md

```diff
@@ -0,0 +1,278 @@
+* To make feasible the interoperability of components provided by different tiers,
+  thin-edge comes with an extensible data model for IIoT -
+  cloud connectivity, telemetry data, device management, child devices, ...
+
+## Design Principles
+
+* Thin-edge provides the tools to develop an IIoT agent by assembling components
+  that have been implemented independently,
+  possibly by different vendors and using different programming languages.
+* There are two levels of components:
+  * MQTT-based components that are processes running on the local network
+    and interacting over MQTT and HTTP using JSON messages.
+  * Rust-based components that are actors running inside a process
+    and interacting over in-memory channels using statically typed messages.
+* These two levels of components serve different purposes and users:
+  * The Rust-based components and their assemblage are designed for robustness.
+    They follow strict combination rules enforced at compile time,
+    notably with compatibility checks between message consumers and producers.
+    These components provide the building blocks:
+    * connections to IoT cloud end-points,
+    * connections to various protocols (MQTT, HTTP, Modbus, OPC UA, ...),
+    * interactions with the operating systems (files, commands, software updates, ...),
+    * making sense of telemetry data (measurements, events, alarms, set points).
+  * The MQTT-based components bring the flexibility to interact
+    with external systems running on a different device or inside a container
+    and without enforcing a programming language or an operating system.
+    These components also provide the flexibility to experiment
+    and to add missing features without having to implement a full-fledged Rust component. 
+* To enable interactions between the MQTT-components, thin-edge provides an MQTT bus made of:
+    * a local MQTT server,
+    * an MQTT bridge that relays messages between the agent and the cloud end-points,
+    * an MQTT API that defines topics and message payloads
+      * to exchange telemetry data,
+      * to monitor components health,
+      * to trigger operations and monitor progress,
+    * a local HTTP server to exchange files.
+* In practice, the software for an IIoT agent is build using a combination of MQTT and Rust components.
+  * Thin-edge itself is released as an MQTT-based component built from Rust-based components.
+  * A batteries included thin-edge executable is available to let the users experiment
+    with all the available building-blocks.
+  * An application-specific thin-edge can be built by cherry-picking building blocks
+    and adding custom blocks for the business logic.
+  * Along the main thin-edge executable, are deployed MQTT-based components over the local network,
+    on external devices, controllers, PLCs, containers ... or even on the gateway
+    connecting all the data sources, actuators, and data processors that make the application on the edge.
```

Comment:
Should we add one more point, which is the "seamless conversion of a Rust-based actor into an independent MQTT component"?

Reply:
Not sure that this can be seamless and even meaningful for all actors. I will try to add a point about that, though.